File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        ans=[]
        ans.append(0)
        for i in range(1,len(prices)):
            ans.append(prices[i]-prices[i-1])

        max_=0
        total=0
        for i in range(len(ans)):

            max_+=ans[i]
            if(max_>total):
                    total=max_;
            if(max_<0):
                    max_=0
            
        return total
        

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        
       
            vector<int>prefix;
            prefix.push_back(0);
            for(int i=1;i<prices.size();i++){
                        prefix.push_back(prices[i]-prices[i-1]); 
                     
            }
            
            int profit=0;
            int max_profit=0;
            for(int i=0;i<prefix.size();i++){
                        
                        profit+=prefix[i];
                        if(profit<0){
                                
                                    profit=0;
                            
                        }
                        
                        max_profit=max(max_profit,profit);
                        
                            
            }
            
            return max_profit;
            
        
        
    }
};